src/dbnode/storage/bootstrap/bootstrapper/base.go

```diff
@@ -76,133 +74,261 @@ func (b baseBootstrapper) String() string {
+		if !nextNamespace.DataRunOptions.ShardTimeRanges.IsEmpty() ||
+			!nextNamespace.IndexRunOptions.ShardTimeRanges.IsEmpty() {
+			next.Namespaces.Set(id, nextNamespace)
+		}
 
-	step.mergeResults(unfulfilled)
+		// Log the result.
+		successLogFields := append(logFieldsCopy(logFields), []zapcore.Field{
+			zap.String("namespace", id.String()),
+			zap.Int("numShards", len(currNamespace.Shards)),
+		}...)
+		successLogFields = append(successLogFields,
+			zap.Duration("dataRangeRequested", dataCurrRequested.MinMaxRange()))
+		successLogFields = append(successLogFields,
+			zap.Duration("dataRangeFulfilled", dataCurrFulfilled.MinMaxRange()))
+		successLogFields = append(successLogFields,
+			zap.Duration("indexRangeRequested", indexCurrRequested.MinMaxRange()))
+		successLogFields = append(successLogFields,
+			zap.Duration("indexRangeFulfilled", indexCurrFulfilled.MinMaxRange()))
+
+		successLogFields = append(successLogFields,
+			zap.Int("dataNumSeries", currResult.DataMetadata.NumSeries))
+		if namespace.Metadata.Options().IndexOptions().Enabled() {
+			successLogFields = append(successLogFields,
+				zap.Int("indexNumSeries", currResult.IndexMetadata.NumSeries))
+		}
+
+		b.log.Info("bootstrapping from source completed successfully",
+			successLogFields...)
 
-	unattemptedNextRanges := currRanges.Copy()
-	unattemptedNextRanges.Subtract(currStatus.fulfilled)
-	if !nextAttempted {
-		// If we have never attempted the next time range then we want to also
-		// attempt the ranges we didn't even try to attempt
-		unattemptedNextRanges.AddRanges(nextRanges)
 	}
 
+	// Unless next bootstrapper is required, this is the final results.
+	finalResults := currResults
+
 	// If there are some time ranges the current bootstrapper could not fulfill,
 	// that we can attempt then pass it along to the next bootstrapper.
-	// NB(r): We explicitly do not ask the next bootstrapper to retry ranges
-	// it could not fulfill as it's unlikely to be able to now.
-	if !unattemptedNextRanges.IsEmpty() {
-		nextStatus, nextErr = step.runNextStep(unattemptedNextRanges)
-		if nextErr != nil {
-			return nextErr
+	if next.Namespaces.Len() > 0 {
+		nextResults, err := b.next.Bootstrap(next)
+		if err != nil {
+			return bootstrap.NamespaceResults{}, err
 		}
 
-		unfulfilledFinal := unfulfilled.Copy()
-		unfulfilledFinal.Subtract(nextStatus.fulfilled)
-		step.mergeResults(unfulfilledFinal)
+		// Now merge the final results.
+		for _, elem := range nextResults.Results.Iter() {
+			id := elem.Key()
+			namespace := elem.Value()
+
+			finalResult, ok := finalResults.Results.Get(id)
+			if !ok {
+				return bootstrap.NamespaceResults{},
+					fmt.Errorf("expected result for namespace: %s", id.String())
+			}
+
+			// NB(r): Since we originally passed all unfulfilled ranges to the
+			// next bootstrapper, the final unfulfilled is simply what it could
+			// not fulfill.
+			finalResult.DataResult.SetUnfulfilled(namespace.DataResult.Unfulfilled().Copy())
```

Comment:
So I'm pretty sure this is fine, but just to double check, if the first bootstrap had unfulfilled results, this will include them and not clobber them right?

Reply:
Correct, this is because the set of ranges we asked the second one to bootstrap including all unfulfilled from the first.  This way whatever the next bootstrapper was unable to unfulfill, is what the entire set of both of them was not able to fulfill.